# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis with pandas. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [99]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Import the appropriate libraries with aliases
# Create two new dataframes
northCC = pd.read_csv(r'C:\Users\willm\Desktop\launchcode\callcenter\call-center-cleanup-assignment\NorthCallCenter.csv')
southCC = pd.read_csv(r'C:\Users\willm\Desktop\launchcode\callcenter\call-center-cleanup-assignment\SouthCallCenter.csv')



In [100]:

pd.set_option('display.max_columns',None)
# pd.set_option('display.max_rows',None)
pd.set_option('display.max_colwidth',None)

display(northCC)
display(southCC)

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0             0  North        Complaint    9:00 AM             Incoming     B   
1             1  North    Sales Support   11:00 AM             Incoming     A   
2             2  North    Sales Support   11:00 AM             Incoming     B   
3             3  North    Sales Support    4:00 PM             Incoming     B   
4             4  North  Product Support   12:00 PM             Incoming     B   
..          ...    ...              ...        ...                  ...   ...   
240         240  North  Product Support    5:00 PM             Outgoing     B   
241         241  North  Product Support    5:00 PM             Outgoing     A   
242         242  North  Product Support    5:00 PM             Outgoing     A   
243         243  North  Product Support    5:00 PM             Outgoing     A   
244         244  North  Product Support    5:00 PM             Outgoing     B   

     Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     Brent   NO          0          1             9                  2   
1       Cam   NO          0          0            11                  1   
2      Todd   NO          0          0             9                  3   
3     Brent   NO          0          0            11                  2   
4     Brent   NO          0          0             8                  2   
..      ...  ...        ...        ...           ...                ...   
240   Brent   NO          0          0            10                  2   
241     Joe   NO          0          0             8                  3   
242  Amanda   NO          0          0            12                  1   
243     Cam   NO          0          0            11                  2   
244  Xander   NO          0          0             8                  1   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      4      1  
1                   1.0                      3      1  
2                   1.0                      3      1  
3                   1.0                      4      1  
4                   1.0                      4      1  
..                  ...                    ...    ...  
240                 NaN                      3      1  
241                 NaN                      2      1  
242                 NaN                      2      1  
243                 NaN                      2      1  
244                 NaN                      4      1  

[245 rows x 15 columns]

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0             0  South    Sales Support   10:00 AM             Incoming     D   
1             1  South    Sales Support   10:00 AM             Incoming     C   
2             2  South    Sales Support   10:00 AM             Incoming     C   
3             3  South    Sales Support   10:00 AM             Incoming     C   
4             4  South    Sales Support   12:00 PM             Incoming     C   
..          ...    ...              ...        ...                  ...   ...   
309         309  South  Product Support    5:00 PM             Outgoing     D   
310         310  South  Product Support    5:00 PM             Outgoing     C   
311         311  South  Product Support    5:00 PM             Outgoing     C   
312         312  South  Product Support    5:00 PM             Outgoing     C   
313         313  South  Product Support    5:00 PM             Outgoing     D   

     Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0      Kate   NO          0          0             9                  1   
1      Eric   NO          0          0             8                  2   
2     Susan   NO          0          0            10                  2   
3     Alice   NO          0          0            12                  1   
4     Sandy   NO          0          0             8                  3   
..      ...  ...        ...        ...           ...                ...   
309   Helga   NO          0          0            10                  3   
310   Susan   NO          0          0            12                  3   
311   Sandy   NO          0          0             9                  3   
312    Eric   NO          0          0            10                  1   
313  George   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      2      1  
1                   1.0                      4      1  
2                   1.0                      4      1  
3                   1.0                      3      1  
4                   1.0                      3      1  
..                  ...                    ...    ...  
309                 NaN                      4      1  
310                 NaN                      4      1  
311                 NaN                      4      1  
312                 NaN                      2      1  
313                 NaN                      3      1  

[314 rows x 15 columns]

## Task 1: Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, answer the following questions. Use the code blocks below to begin cleaning your data. At the end of the section, record your answers.

#### EDA Question 1A:  How many reps are in the North branch?  

In [101]:
# EDA Question 1A solution below:
northCC['Rep ID'].dropna().nunique()

# There are 9 unique Rep IDs in the North Branch.

9

#### EDA Question 1B:  How many reps are in the South branch?  

In [91]:
# EDA Question 1B solution Below:
southCC["Rep ID"].dropna().nunique()

# There are 11 unique Rep IDs in the South Branch

11

#### EDA Question 2A:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the North branch? 

In [102]:
# EDA Question 2B solution Below:
numericnorth = northCC.select_dtypes(include='number')

avgnorth = numericnorth.agg('mean')
display(avgnorth)

# Average 'busy minutes' is 9.99. Average 'not ready minutes' is 1.91.
# Average 'incoming wait time' is 3.05 minutes. Average 'during call wait time' is 2.97 minutes.


Unnamed: 0               122.000000
Lost Call                  0.020408
Abandoned                  0.028571
Busy Minutes               9.987755
Not Ready Minutes          1.914286
Incoming Wait Time         3.048780
During Call Wait Time      2.967347
Calls                      1.000000
dtype: float64

#### EDA Question 2B:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the South branch? 

In [103]:
# EDA Question 2B solution Below:
numericsouth = southCC.select_dtypes(include='number')

avgsouth = numericsouth.agg('mean')
display(avgsouth)

# Average 'busy minutes' is 10.05. Average 'not ready minutes' is 1.91.
# Average 'incoming wait time' is 3.00 minutes. Average 'during call wait time' is 3.08 minutes.


Unnamed: 0               156.500000
Lost Call                  0.035032
Abandoned                  0.012739
Busy Minutes              10.054140
Not Ready Minutes          1.910828
Incoming Wait Time         3.000000
During Call Wait Time      3.076433
Calls                      1.000000
dtype: float64

#### EDA Question 3:  What is the number of calls taken for each time block(Both North and South combined)?

In [104]:
# EDA Question 3 solution Below:

groupednorth = northCC.groupby('Time Block')['Calls'].sum()

groupedsouth = southCC.groupby('Time Block')['Calls'].sum()

groupedtotal = groupednorth.add(groupedsouth, fill_value=0)

print(groupedtotal)


Time Block
10:00 AM     99
11:00 AM     56
12:00 PM    120
1:00 PM      40
2:00 PM      65
3:00 PM      73
4:00 PM      53
5:00 PM      43
9:00 AM      10
Name: Calls, dtype: int64


##### Record your answers for the EDA section below:
- 1a) Reps in the North Branch = 9
- 1b) Reps in the South Branch = 11


- 2a) North Branch, (round to the nearest hundredth):
> - Busy Minutes = 9.99
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.05
> - During Call Wait Time = 2.97
- 2b) South Branch, (round to the nearest hundredth):
> - Busy Minutes = 10.05
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.00
> - During Call Wait Time = 3.08
- 3) Total Calls taken for all branches per time block:
> - 9:00AM   = 10
> - 10:00AM  = 99
> - 11:00AM  = 56
> - 12:00PM  = 120
> - 1:00PM   = 40
> - 2:00PM   = 65
> - 3:00PM   = 73
> - 4:00PM   = 53
> - 5:00PM   = 43



## Task 2: Cleaning Your Data 
Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. 

#### Cleaning Question 1:  Is there any missing data and if yes explain what you would do with the missing data and why?

In [63]:
# Question 1 solution below

display(northCC.isna().sum())
display(southCC.isna().sum())



Unnamed: 0                 0
Branch                     0
Call Purpose               0
Time Block                 0
Incoming or Outgoing       0
Queue                      0
Rep ID                     0
Sale                       0
Lost Call                  0
Abandoned                  0
Busy Minutes               0
Not Ready Minutes          0
Incoming Wait Time       163
During Call Wait Time      0
Calls                      0
dtype: int64

Unnamed: 0                 0
Branch                     0
Call Purpose               0
Time Block                 0
Incoming or Outgoing       0
Queue                      0
Rep ID                     0
Sale                       0
Lost Call                  0
Abandoned                  0
Busy Minutes               0
Not Ready Minutes          0
Incoming Wait Time       188
During Call Wait Time      0
Calls                      0
dtype: int64

#### Record your answer below:

> Your Answer: There is data missing in the Incoming Wait Time column. 

#### Cleaning Question 2:  In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?

In [105]:
# Cleaning Question 2 solution below:

print(northCC['Sale'].unique())


['NO' 'YES ' 'YES']


In [106]:
northCC['Sale'].str.rstrip()

0      NO
1      NO
2      NO
3      NO
4      NO
       ..
240    NO
241    NO
242    NO
243    NO
244    NO
Name: Sale, Length: 245, dtype: object

##### Record your answer by below:
> Your Answer: 

#### Cleaning Question 3:  Are there duplicates in the two data sets? If there are, how did you handle them?

In [107]:
# Cleaning Question 3 solution below:

northdupes = northCC[northCC.duplicated()]
southdupes = southCC[southCC.duplicated()]

print(northdupes)
print(southdupes)

Empty DataFrame
Columns: [Unnamed: 0, Branch, Call Purpose, Time Block, Incoming or Outgoing, Queue, Rep ID, Sale, Lost Call, Abandoned, Busy Minutes, Not Ready Minutes, Incoming Wait Time, During Call Wait Time, Calls]
Index: []
Empty DataFrame
Columns: [Unnamed: 0, Branch, Call Purpose, Time Block, Incoming or Outgoing, Queue, Rep ID, Sale, Lost Call, Abandoned, Busy Minutes, Not Ready Minutes, Incoming Wait Time, During Call Wait Time, Calls]
Index: []


##### Record your answer below:
> Your Answer: There are no duplicates, but if there had been I would have used the .drop_duplicates() function on each dataframe.

#### Cleaning Question 4:  Is any of the data in the two data sets unnecessary? If yes, how did you handle it?

In [108]:
# Cleaning Question 4 solution below:
northCC.drop(columns=['Unnamed: 0'], inplace=True)
southCC.drop(columns=['Unnamed: 0'], inplace=True)

##### Record your answer below:
> Your Answer: 'Unnamed: 0' was just a duplicate of the index, and 'Calls' only contained 1s. I dropped them from the dataset by using the .drop function.

## Task 3: Data Manipulation
Before you begin answering any questions, combine the two datasets together to create a third dataframe called df. You can use this third dataframe to compare the two call centers to company-wide trends.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. Record your answer below.

In [113]:
# Create dataframe for the entire company named df

df = pd.concat([northCC,southCC])
df = df.drop_duplicates()
df.reset_index(drop=True)

print(df)

    Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    North        Complaint    9:00 AM             Incoming     B   Brent   
1    North    Sales Support   11:00 AM             Incoming     A     Cam   
2    North    Sales Support   11:00 AM             Incoming     B    Todd   
3    North    Sales Support    4:00 PM             Incoming     B   Brent   
4    North  Product Support   12:00 PM             Incoming     B   Brent   
..     ...              ...        ...                  ...   ...     ...   
309  South  Product Support    5:00 PM             Outgoing     D   Helga   
310  South  Product Support    5:00 PM             Outgoing     C   Susan   
311  South  Product Support    5:00 PM             Outgoing     C   Sandy   
312  South  Product Support    5:00 PM             Outgoing     C    Eric   
313  South  Product Support    5:00 PM             Outgoing     D  George   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO

#### Manipulation Question 1:  Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.

In [112]:
# Manipulation Question solution below:
print(df.duplicated().sum())
# grouped = df.groupby(['Branch', 'Rep ID']).count()
# print(grouped)

5


##### Record your answer below
Rep with the hightest number of calls and their total calls:
- North Branch = Brent (37)
- South Branch = Eric (35)

#### Manipulation Question 2:  The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?

In [80]:
# Manipulation Question 2 solution below:



    Branch     Call Purpose Time Block Incoming or Outgoing Queue Rep ID Sale  \
253  South  Product Support    2:00 PM             Incoming     D  Randy   NO   
259  South  Product Support    5:00 PM             Incoming     D  Randy   NO   
260  South        Complaint   11:00 AM             Incoming     D  Randy   NO   
265  South        Complaint   12:00 PM             Incoming     D  Randy   NO   
267  South    Sales Support   10:00 AM             Incoming     D  Randy   NO   
270  South    Sales Support    1:00 PM             Incoming     D  Randy   NO   
279  South  Product Support    3:00 PM             Incoming     D  Randy   NO   
313  South  Product Support    3:00 PM             Incoming     D  Randy   NO   
314  South  Product Support    3:00 PM             Incoming     D  Randy   NO   
316  South  Product Support    5:00 PM             Incoming     D  Randy   NO   
322  South        Complaint    4:00 PM             Incoming     D  Randy   NO   
323  South    Sales Support 

##### Record your answer below:
Average Call Abandonment Rates (round to the nearest hundredth):
- North Branch = 
- South Branch = 
- Company Wide =
- Do any of these fall out of the optimal range of 2-5%?

#### Manipulation Question 3:  Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?

In [54]:
# Manipulation Question 3 solution below:

##### Record your answer below:
Percentage of calls answered within 2 seconds, include # of calls:
- North Branch = 
- South Branch = 
- Company Wide = 

#### Manipulation Question 4: For each branch and the entire company, what is the average speed of answer?

In [55]:
# Manipulation Question 4 solution below:

##### Record your answer by below:
Average speed of answer (rounded to nearest hundredth):
- North Branch in seconds = 
- South Branch in seconds = 
- Company Wide in seconds = 

## Task 4: Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why. 

*NOTE Some questions you may decide to use more than one chart and or chart style.

#### Visualization 1:  What is the average abandonment rate per queue?

In [56]:
# Create visualization 1 here
# Import any additional libraries needed with alias

# The dictionary abandonment_rates has the data you need.
abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

#Your code below:


AttributeError: 'NoneType' object has no attribute 'groupby'

##### Record your answer below:

> Chart style you chose and why: 

#### Visualization 2: What is the service level percentage and average speed of answer for each rep in the North Branch?

In [ ]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

#Your Code Here:


##### Record your answer below:

> Chart style you chose and why: 

#### Visualization 3: For each type of call purpose, how many calls are outgoing vs. incoming?

In [ ]:
# Create visualization 3 here:
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_group = df.groupby("Call Purpose")
call_purpose = ["Complaint", "Product Support", "Sales Support"]
purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
print(purpose_counts)

complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

#Your Code Here:


##### Record your answer below:

> Chart style you chose and why: 

## Task 5: Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.


#### Summary Question 1:  Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?

##### Record your answer below:
> Your Answer: 



#### Summary Question 2: Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?

##### Record your answer below:
> Your Answer: 


#### Summary Question 3: Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

##### Record your answer below:
> Your Answer: 


## Bonus Mission
Create a visualization that answers this question: For each call purpose, how many calls (incoming and outgoing) take place in each time block?
##### Record your answer below:

> Chart style you chose and why: 

In [ ]:
# Create your Bonus Mission visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot: